theseus/core/loss.py

```diff
@@ -0,0 +1,42 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import abc
+
+import torch
+
+
+class Loss:
+    @staticmethod
+    @abc.abstractmethod
+    def evaluate(x: torch.Tensor, radius: torch.Tensor) -> torch.Tensor:
+        pass
+
+    @staticmethod
+    @abc.abstractmethod
+    def linearize(x: torch.Tensor, raidus: torch.Tensor) -> torch.Tensor:
+        pass
+
+
+class WelschLoss(Loss):
+    @staticmethod
+    def evaluate(x: torch.Tensor, radius: torch.Tensor) -> torch.Tensor:
+        return radius - radius * torch.exp(-x / radius)
```

Comment:
Not sure if this is needed but do we raise a warning if radius is zero/close to zero. Incase you are learning it, there is a chance this fails silently by sending nan/infs.